earthfile2llb/converter.go

```diff
@@ -1300,8 +1300,16 @@ func (c *Converter) FinalizeStates(ctx context.Context) (*states.MultiTarget, er
 }
 
 // ExpandArgs expands args in the provided word.
-func (c *Converter) ExpandArgs(word string) string {
-	return c.varCollection.Expand(word)
+func (c *Converter) ExpandArgs(ctx context.Context, word string) (string, error) {
+	if !variables.ContainsShell(word) {
+		return c.varCollection.Expand(word), nil
+	}
+	pncvf := c.processNonConstantBuildArgFunc(ctx)
```

Comment:
Another situation: what if there are multiple parts in the string to replace.

```
ARG greeting="hello $(id -F), the date is $(date)"
```

Reply:
I'll create another test to verify this two-parter works.

Reply:
this handles multiple parts plus an env:
https://github.com/earthly/earthly/pull/1699/files#diff-b3414e5141e314866d254a961baf56400082c346f72c0633d94b3ec745dfd93bR85-R91